### ISSUE TITLE
Full state saving corrupts ui:repeat and UIData: SKIP_ITERATION is ignored, and rowStatePreserved never resets row client IDs

### ISSUE THREAD

**Issue by BalusC** [eclipse-ee4j/mojarra#5927]
Discovered while running 5.0 TCK with `-Dwebapp.partialStateSaving=false` since context param improvements in https://github.com/eclipse-ee4j/mojarra/issues/5903.

Two defects in row-scoped client ID handling. Both are invisible under partial state saving and both break the row-less/row-scoped client ID contract that full state saving depends on, so they are reported together.

## 1. `UIRepeat` ignores `SKIP_ITERATION` when the behavior source is inside it

Under full state saving, any `ui:repeat` containing the ajax source loses **all** component state for its children on postback, and the render-time facelet re-apply then duplicates the entire subtree.

`UIRepeat.requiresRowIteration(VisitContext)` overrides the caller's `SKIP_ITERATION` hint whenever the request's `jakarta.faces.source` starts with the repeat's client ID. `FaceletFullStateManagementStrategy.saveComponentState()` and `restoreComponentState()` both visit with `SKIP_ITERATION` and key the saved state by client ID, so they depend on that hint being honored. On the initial GET the override is inactive and state is filed under `form:repeat:checkbox`; on the ajax postback the override fires and the restore visit looks up `form:repeat:0:checkbox`, which is not in the state map. Every component inside the repeat is therefore restored with no state at all — including its `facelets.MARK_ID` — so `ComponentSupport.findChildByTagId` cannot match it during the render-time re-apply and the tag handlers append a second copy.

Instrumented trace of one ajax postback:

```
@@@SAVEKEY@@@    j_idt6:j_idt8:checkbox    HtmlSelectBooleanCheckbox
@@@RESTOREKEY@@@ j_idt6:j_idt8:0:checkbox  HtmlSelectBooleanCheckbox hasState=false
@@@APPLY@@@ tagId=…3db77da parent=HtmlForm/j_idt6  found=true   tag=ui:repeat
@@@APPLY@@@ tagId=…3db7734 parent=UIRepeat/j_idt8 found=false  tag=h:selectBooleanCheckbox
```

`jakarta.faces.component.UIData.requiresRowIteration` carries no such override — it is exactly `return !ctx.getHints().contains(VisitHint.SKIP_ITERATION);`. Aligning `UIRepeat` with the spec component fixes it:

```java
 private boolean requiresRowIteration(VisitContext ctx) {
-    boolean shouldIterate = !ctx.getHints().contains(VisitHint.SKIP_ITERATION);
-    if (!shouldIterate) {
-        FacesContext faces = ctx.getFacesContext();
-        String sourceId = BEHAVIOR_SOURCE_PARAM.getValue(faces);
-        boolean containsSource = sourceId != null ? sourceId.startsWith(super.getClientId(faces) + getSeparatorChar(faces)) : false;
-        return containsSource;
-    } else {
-        return shouldIterate;
-    }
+    return !ctx.getHints().contains(VisitHint.SKIP_ITERATION);
 }
```

## 2. `rowStatePreserved` renders every row under row 0's client ID

With `rowStatePreserved="true"` and full state saving, every row of a `ui:repeat` or `h:dataTable` renders its children under row 0's client ID. This shows on the plain initial GET, before any postback:

```html
<input id="form:outer:0:inner:0:input"><input id="form:outer:0:inner:0:input">
<input id="form:outer:0:inner:0:input">  <!-- all six cells of a 2x3 matrix -->
```

`UIRepeat.markInitialState()` and `UIData.markInitialState()` capture `initialDescendantFullComponentState` only when `StateManager.IS_BUILDING_INITIAL_STATE` is present. Full state saving never sets it, because `FaceletViewHandlingStrategy.markInitialState(ctx, view)` is gated on partial state saving, and the field is not part of `saveState` either. Both branches of `setRowIndexRowStatePreserved` / `setRowIndexWithRowStatePreserved` are then skipped — and the per-row client ID reset, `component.setId(component.getId())` in `restoreFullDescendantComponentStates` (commented *"reset the client id (see spec 3.1.6)"*), lives inside that same skipped branch.

Preserving full per-row state is inherently a partial-state-saving construct, so it cannot be honored here — but it must degrade to the ordinary per-row submitted-value state rather than stop managing client IDs. Select the branch on whether the state it needs is actually available:

```java
 public void setRowIndex(int rowIndex) {
-    if (isRowStatePreserved()) {
+    if (isRowStatePreserved() && _initialDescendantFullComponentState != null) {
         setRowIndexRowStatePreserved(rowIndex);
     } else {
         setRowIndexWithoutRowStatePreserved(rowIndex);
     }
 }
```

The same change applies to `UIRepeat.setIndex`. Note the `UIData` half lives in `jakarta.faces.component.UIData` in the API repository, so this defect spans both repositories.

Unlike defect 1 this one is not masked by the id-uniqueness walk: it reproduces in Development stage too.

## Affected branches

`requiresRowIteration` is byte-identical on `4.0`, `4.1` and `master`, and dates to the initial Oracle contribution in 2018 — it is not a regression. What changed is only whether it is observable:

| branch | `disableIdUniquenessCheck` default | full state saving |
|---|---|---|
| 4.0 | `"false"` — always check | masked |
| 4.1 | `"false"` — always check | masked |
| master (5.0) | `auto` — skipped unless Development | **exposed** since 28c663f31e |

`Util.checkIdUniqueness` walks the whole tree calling `getClientId()` on every node during `saveView`, which keeps the client ID key space row-less and accidentally hides defect 1. 4.0 and 4.1 still run that walk by default, but reach the bug as soon as `disableIdUniquenessCheck=true` is set, which is a documented performance opt-in.

Defect 2 is structural wherever the feature exists: `UIData.rowStatePreserved` long-standing, `UIRepeat.rowStatePreserved` since 4.1.

## Verification

Jakarta Faces TCK with `-Dwebapp.partialStateSaving=false`.

Fixed by defect 1's change: `Issue1821IT`, `Issue2640IT`, `Issue2910IT` (faces22/ajax-lifecycle), `Issue1581IT`, `Issue5078IT` (faces23/facelets).

Fixed by defect 2's change: `NestedRepeatRepeatRowStatePreservedIT`, `NestedTableTableRowStatePreservedIT` (faces23/faces-data-model).

Fully green under full state saving with both: faces22/ajax-lifecycle 22/22, faces23/faces-data-model 25/25, faces41/api 18/18, faces50/facelets 33/33. No partial-state-saving regressions across 250 tests in the 12 TCK modules that combine `ui:repeat` or `h:dataTable` with ajax.

🤖 Generated with Claude Opus 5

**Comment by BalusC** [comment]
> and dates to the initial Oracle contribution in 2018

We need to doublecheck if this doesn't regress https://github.com/eclipse-ee4j/mojarra/issues/3837. That `containsSource` check was added there. Next time please dig further in https://github.com/javaee/mojarra to find original issues.

**Comment by BalusC** [comment]
The javaee/mojarra history gives the origin: 42a30d96bd, 2015-05-20, for JAVASERVERFACES-3833 ("UIRepeat, Ajax and state"), authored by Michal Petrov of Red Hat and reviewed by Ed Burns (svn trunk@14764), backported the next day as 68341f55d3 for JAVASERVERFACES-3936. That JAVASERVERFACES-3833 on JIRA is what later became #3837, so the check is exactly the one added there.

Before that commit the method read, byte for byte, what this issue proposes restoring, so the change is a literal revert of it:

```java
private boolean requiresRowIteration(VisitContext ctx) {
    return !ctx.getHints().contains(VisitHint.SKIP_ITERATION);
}
```

The commit shipped its own regression test — `Issue3833IT` plus `issue3833.xhtml` and `Issue3833Bean` — which lives in the TCK today as `Issue3837IT` in faces22/ajax-inputs, same page and same assertions: a `ui:repeat` of `h:selectOneListbox` with a `valueChangeListener` and `f:ajax`, asserting each row's own previous value across four ajax postbacks (`null -> 1`, `null -> 3`, `1 -> 2`, `3 -> 4`).

**It does not regress.** With the check removed, `Issue3837IT` passes under both partial and full state saving, and faces22/ajax-inputs is 41/41 under partial state saving.

One caveat on reproducing that: `Issue3837IT` is `@Disabled`, following the accepted challenge in jakartaee/faces#1757 which excluded it on 2023-01-24. A normal TCK run therefore does not exercise this scenario at all, so it has to be enabled explicitly. Anyone re-validating this change should do that rather than rely on a full TCK run.

`UIData` has never carried an equivalent check, and covers the same scenario by saving and restoring its descendants' per-row state around each row change, in `saveDescendantState`/`restoreDescendantState` from `setRowIndexWithoutRowStatePreserved`. Which change on the `UIRepeat` side made the 2015 workaround unnecessary is not yet identified; what is established is that its own regression test no longer depends on it.

There is a side effect for #3156 as well. `requiresRowIteration` returning `true` makes `visitTree` and `doVisitChildren` call `setIndex(ctx, -1)`, which builds the `DataModel` and so evaluates `value` — during a state-saving visit, for any repeat that happens to contain the ajax source. Dropping the check removes those forced builds. `Issue3156IT` stays green with the change under both state saving modes.

🤖 Generated with Claude Opus 5

**Comment by BalusC** [comment]
It was most probably solved by one of performance improvements done in #5753 whereby much of UIRepeat was (re)aligned with UIData.

Issue3837IT can be reenabled after https://github.com/jakartaee/faces/issues/2206 is addressed.
